File: scripts/youngcha/producers/sporefarm.js

const crucibleLib = require("libraries/cruciblelib");
//tile.floor().liquidDrop == Liquids.water

const frames = 5;
const tilechkdirs = [
	{x:-1,y: 1},{x:0,y: 1},{x:1,y: 1},
	{x:-1,y: 0},/*[tile]*/ {x:1,y: 0},
	{x:-1,y:-1},{x:0,y:-1},{x:1,y:-1},
];

const sporeFarm = extendContent(Block, "spore-farm", {
	_timerid:0,
	_dumptimerid:0,
	load(){
		this.super$load();
		this.plantSprite = [Core.atlas.find(this.name+"-spore1"),Core.atlas.find(this.name+"-spore2"),Core.atlas.find(this.name+"-spore3"),Core.atlas.find(this.name+"-spore4"),Core.atlas.find(this.name+"-spore5")];
		this.bottom  = [Core.atlas.find(this.name+"-ground1"),Core.atlas.find(this.name+"-ground2"),Core.atlas.find(this.name+"-ground3"),Core.atlas.find(this.name+"-ground4"),Core.atlas.find(this.name+"-ground5")];
		this.fence  = Core.atlas.find(this.name+"-fence");
	},
	getTimerId(){
		return this._timerid;
	},
	getDTimerId(){
		return this._dumptimerid;
	}
});

const gtimer = sporeFarm.timers++;
const dtimer = sporeFarm.timers++;

sporeFarm.buildType = () => 
{
	let building = extend(Building, {
		growth:0,
		delay:-1,
		tileIndex:-1,
		needsTileUpdate: false,
		triggerTileUpdate(){
			this.needsTileUpdate=true;
		},
		randomChk() {
			let ctile = Vars.world.tile(Math.floor(this.tile.x+Mathf.range(3)),Math.floor(this.tile.y+Mathf.range(3)));
			return ctile && ctile.floor().liquidDrop == Liquids.water;
		},
		updateTilings(){
			let bitmask = 0;
			for(let i =0;i<tilechkdirs.length;i++){
				let tile = this.tile.nearby(tilechkdirs[i].x, tilechkdirs[i].y);
				if(!tile || !tile.bc() || !tile.bc().updateTilings){
					continue;
				}
				bitmask += 1<<i;
			}
			this.tileIndex=bitmask;
		},
		updateNeighbours(){
			for(let i =0;i<tilechkdirs.length;i++){
				let tile = this.tile.nearby(tilechkdirs[i].x, tilechkdirs[i].y);
				if(!tile || !tile.bc() || !tile.bc().triggerTileUpdate){
					continue;
				}
				tile.bc().triggerTileUpdate();
			}
		},
		onProximityRemoved() {
			this.super$onProximityRemoved();
			this.updateNeighbours();
		},
		updateTile(){
			if(this.tileIndex==-1){
				this.updateTilings();
				this.updateNeighbours();
			}
			if(this.needsTileUpdate){
				this.updateTilings();
				this.needsTileUpdate = false;
			}
			if(this.timer.get(gtimer, (60+this.delay)*5.0 )){
				if(this.delay==-1){
					this.delay =  (this.tile.x*89+this.tile.y*13)%21;
				}else{		
					if(this.growth==0 && (!this.randomChk())){
						return;
					}
					this.growth+=this.randomChk()?(this.growth>frames-2?0.1:0.45):-0.1;
					if(this.growth>=frames){
						this.growth=frames-1;
						if(this.items.total()<1){
							this.offload(Items.sporePod);
						}
					}
					if(this.growth<0){
						this.growth=0;
					}
				}
			}
			if(this.timer.get(dtimer,15)){
				this.dump(Items.sporePod);
			}
		},
		draw() {
			let rrot =  (this.tile.x*89+this.tile.y*13)%4;
			let rrot2 =  (this.tile.x*69+this.tile.y*42)%4;
			if(this.growth!=0){
				Draw.rect(sporeFarm.bottom[Math.floor(this.growth)], this.x, this.y, rrot*90.0);
				Draw.rect(sporeFarm.plantSprite[Math.floor(this.growth)], this.x, this.y, rrot2*90.0);
			}
			Draw.rect(crucibleLib.getRegion(sporeFarm.fence, crucibleLib.tileIndexMap[this.tileIndex]), this.x, this.y, 8, 8, 0);
			this.drawTeamTop();
		},
		write(stream) {
			this.super$write(stream);
			stream.f(this.growth);
		},
		read(stream, revision) {
			this.super$read(stream,revision);
			this.growth=stream.f();
		}
	});
	building.block = sporeFarm;
	return building;
};
sporeFarm.update = true;
sporeFarm.solid = false;

File: main/src/unity/world/blocks/production/SporeFarm.java
package unity.world.blocks.production;

import arc.graphics.g2d.*;
import arc.math.*;
import arc.math.geom.*;
import arc.util.io.*;
import mindustry.*;
import mindustry.content.*;
import mindustry.gen.*;
import mindustry.world.*;
import mindustry.world.blocks.environment.*;
import unity.graphics.*;
import unity.util.*;

import static arc.Core.atlas;

public class SporeFarm extends Block{
    static final int frames = 5;
    int gTimer;
    
    public final TextureRegion[] sporeRegions = new TextureRegion[frames], groundRegions = new TextureRegion[frames];
    public TextureRegion[] fenceRegions;
    public TextureRegion cageFloor;

    public SporeFarm(String name){
        super(name);
        
        update = true;
        gTimer = timers++;
    }

    @Override
    public void load(){
        super.load();
        
        for(int i = 0; i < 5; i++){
            sporeRegions[i] = atlas.find(name + "-spore" + (i + 1));
            groundRegions[i] = atlas.find(name + "-ground" + (i + 1));
        }
        
        fenceRegions = Utils.getRegions(atlas.find(name + "-fence"), 12, 4);
        cageFloor = atlas.find(name + "-floor");
    }

    public class SporeFarmBuild extends Building{
        float growth, delay = -1;
        int tileIndex = -1;
        boolean needsTileUpdate;

        boolean randomChk(){
            var cTile = Vars.world.tile(tileX() + Mathf.range(3), tileY() + Mathf.range(3));
            
            return cTile != null && cTile.floor().liquidDrop == Liquids.water;
        }

        void updateTilings(){
            tileIndex = 0;
            
            for(int i = 0; i < 8; i++){
                var other = tile.nearby(Geometry.d8(i));
                
                if(other == null || !(other.build instanceof SporeFarmBuild)) continue;
                tileIndex += 1 << i;
            }
        }

        void updateNeighbours(){
            for(int i = 0; i < 8; i++){
                var other = tile.nearby(Geometry.d8(i));
                
                if(other == null || !(other.build instanceof SporeFarmBuild b)) continue;
                b.needsTileUpdate = true;
            }
        }

        @Override
        public void onProximityRemoved(){
            super.onProximityRemoved();
            
            updateNeighbours();
        }

        @Override
        public void updateTile(){
            if(tileIndex == -1){
                updateTilings();
                updateNeighbours();
            }
            if(needsTileUpdate){
                updateTilings();
                needsTileUpdate = false;
            }
            if(timer(gTimer, (60f + delay) * 5f)){
                if(delay == -1){
                    delay = (tileX() * 89f + tileY() * 13f) % 21f;
                }else{
                    boolean chk = randomChk();
                    
                    if(growth == 0f && !chk) return;
                    growth += chk ? growth > frames - 2 ? 0.1f : 0.45f : -0.1f;
                    
                    if(growth >= frames){
                        growth = frames - 1f;
                        if(items.total() < 1) offload(Items.sporePod);
                    }
                    if(growth < 0f) growth = 0f;
                }
            }
            if(timer(timerDump, 15f)) dump(Items.sporePod);
        }

        @Override
        public void draw(){
            float rrot = (tileX() * 89f + tileY() * 13f) % 4f;
            float rrot2 = (tileX() * 69f + tileY() * 42f) % 4f;
            
            if(growth < frames - 0.5f){
                Tile t = Vars.world.tileWorld(x, y);
                
                if(t != null && t.floor() != Blocks.air){
                    Floor f = t.floor();
                    
                    Mathf.rand.setSeed(t.pos());
                    Draw.rect(f.variantRegions()[Mathf.randomSeed(t.pos(), 0, Math.max(0, variantRegions().length - 1))], x, y);
                }
                
                Draw.rect(cageFloor, x, y);
            }

            if(growth != 0f){
                Draw.rect(groundRegions[Mathf.floor(growth)], x, y, rrot * 90f);
                Draw.rect(sporeRegions[Mathf.floor(growth)], x, y, rrot2 * 90f);
            }
            
            Draw.rect(fenceRegions[UnityDrawf.tileMap[tileIndex]], x, y, 8f, 8f);
            drawTeamTop();
        }

        @Override
        public void write(Writes write){
            super.write(write);
            
            write.f(growth);
        }

        @Override
        public void read(Reads read, byte revision){
            super.read(read, revision);
            
            growth = read.f();
        }
    }
}
